Improve search

bennyz opened the pull request:
Now results are stored in a map which is then iterated and shown
in a satisfactory manner, example:
<pre>
repo: CQRSGolang
        https://api.github.com/repos/TechMaster/CQRSGolang
        stars: 0
repo: go-laboratory
        https://api.github.com/repos/liunian1004/go-laboratory
        stars: 0
</pre>

bennyz committed:
Improve search

Now results are stored in a map which is then iterated and shown
in a satisfactory manner, example:
repo: CQRSGolang
        https://api.github.com/repos/TechMaster/CQRSGolang
        stars: 0
repo: go-laboratory
        https://api.github.com/repos/liunian1004/go-laboratory
        stars: 0

matanshukry commented on backend/rest/rest.go:
```diff
@@ -36,49 +36,61 @@ func New(token string, resultsPerPage int, storageProvider persistence.Storage)
+	repos := make(map[int64]*backends.Result)
+
 	if err != nil {
 		log.Fatal("Failed to fetch results", err)
```
break function flow after this line?

matanshukry commented on backend/rest/rest.go:
```diff
@@ -36,49 +36,61 @@ func New(token string, resultsPerPage int, storageProvider persistence.Storage)
+	repos := make(map[int64]*backends.Result)
+
 	if err != nil {
 		log.Fatal("Failed to fetch results", err)
```
break function flow after this line?

bennyz replied:
Fatal panics and exits

matanshukry replied:
exists app rather than function I'm assuming? is that the desired outcome?

bennyz replied:
no point keeping it running if we didn't get anything, no?

matanshukry commented on backend/rest/rest.go:
```diff
@@ -36,49 +36,61 @@ func New(token string, resultsPerPage int, storageProvider persistence.Storage)
-		fmt.Printf("Fetching value for: %v", repoID)
+	for _, codeResult := range results.CodeResults {
+		repoID := codeResult.GetRepository().GetID()
+		fmt.Printf("Fetching value for: %v\n", repoID)
```
is there any log.debug or something like that?

bennyz replied:
Nope, created issue #2 to address this

matanshukry commented on backend/rest/rest.go:
```diff
@@ -36,49 +36,61 @@ func New(token string, resultsPerPage int, storageProvider persistence.Storage)
-		if err != nil || value == "" {
-			log.Printf("Value not found in DB, requesting amount of stars...\n")
+		if err != nil || value == nil {
 			repo, _, err := c.Client.Repositories.GetByID(c.ctx, repoID)
```
Don't you want to log the error before overriding it?

bennyz replied:
it's logged in db.go, but yeah, I should

matanshukry commented on backend/rest/rest.go:
```diff
@@ -36,49 +36,61 @@ func New(token string, resultsPerPage int, storageProvider persistence.Storage)
-		repos = append(repos, result.GetRepository().GetHTMLURL())
-		repoID := result.GetRepository().GetID()
-		fmt.Printf("Fetching value for: %v", repoID)
+	for _, codeResult := range results.CodeResults {
```
Can you do a multi-get for all the ids from storage?

bennyz replied:
what do you mean?

matanshukry replied:
You're doing:
foreach (repo) { handleResult(storage.get(repo.id)); }

I suggest:
ids = repos.map(repo => repo.id).toList();
results = storage.get(ids);
handleResults(results);

bennyz replied:
yeah, good idea

bennyz committed:
Add sqlite db file to gitignore

bennyz committed:
Add example-finder to .gitignore

bennyz committed:
Optimize search for cached repos

This patch optimizes the DB search for
cached repos by quering for all repo IDs
at once

matanshukry commented on persistence/storage.go:
```diff
@@ -1,7 +1,12 @@
 type Storage interface {
-	Save(key int64, value []byte) (int64, error)
-	Get(key int64) ([]byte, error)
+	Save(key int64, value JSONValue) (int64, error)
```
What is the meaning of the int64 return value? document it

bennyz replied:
I'll remove it

matanshukry commented on util/util.go:
```diff
@@ -0,0 +1,33 @@
+
+// Difference returns the difference between two arrays
+func Difference(arr1, arr2 []int64) (result []int64) {
+	hash := make(map[int64]bool)
```
What happened to the empty struct idea? why not?

bennyz replied:
Will do

matanshukry commented on util/util_test.go:
```diff
@@ -0,0 +1,17 @@
+	"testing"
+)
+
+func TestIntersection(t *testing.T) {
```
Rename to TestDifference

bennyz replied:
Done

matanshukry commented on util/util.go:
```diff
@@ -0,0 +1,33 @@
+)
+
+// Difference returns the difference between two arrays
+func Difference(arr1, arr2 []int64) (result []int64) {
```
Why does the second argument is int64[] and the first has no declaration?

bennyz replied:
I learned a new trick :)
you can specify the type only once if all arguments are of the same type

matanshukry commented on util/util.go:
```diff
@@ -0,0 +1,33 @@
+	hash := make(map[int64]bool)
+
+	for _, item := range arr1 {
+		hash[item] = true
```
Hashing of an "object" - any considerations that should be taken care of?

bennyz replied:
What do you mean?

matanshukry commented on persistence/sqlite/db.go:
```diff
@@ -53,18 +55,16 @@ func (s *sqlite) Save(key int64, value []byte) (int64, error) {
-	} else {
-		log.Printf("Fetched for key %v", key)
+	var value []persistence.JSONValue
+	err = stmt.QueryRow(util.SliceToString(keys)).Scan(&value)
```
Any chance we can avoid SliceToString()? That is, simply pass an array as the variable?

bennyz replied:
I tried, it seems the driver cannot really handle arrays

matanshukry commented on util/util.go:
```diff
@@ -0,0 +1,33 @@
+}
+
+// SliceToString converts an int64 array to a comma separated string
+func SliceToString(arr []int64) string {
```
I'm not a fan of *Slice*ToString. It doesn't really slice anything, does it?

bennyz replied:
Slice is the name of the data structure[1]

[1] - https://tour.golang.org/moretypes/7

matanshukry commented on backend/rest/rest.go:
```diff
@@ -35,53 +38,56 @@ func New(token string, resultsPerPage int, storageProvider persistence.Storage)
 
-// Search searches code using the github api
-func (c *client) Search(query string, lang string) map[int64]*backends.Result {
+var repos map[int64]*backends.Result
```
This is a file-scoped variable, yet it's between 2 functions. Is that standard way in GO?

bennyz replied:
done

matanshukry commented on backend/rest/rest.go:
```diff
@@ -35,53 +38,56 @@ func New(token string, resultsPerPage int, storageProvider persistence.Storage)
+		fmt.Println(err)
+	}
 
+	cachedRepoIds := initializeCache(values)
```
So the entire cache is to cache the calls to GitHub - what about memory cache to sqlite? should we even do it?

bennyz replied:
this is the memory cache, no?
I don't know if something more complex is required given this is a short-living program

matanshukry commented on backend/rest/rest.go:
```diff
@@ -20,6 +22,7 @@ type client struct {
 }
 
 var storage persistence.Storage
+var cachedRepos []*backends.Result
```
am I missing something or are you just populating this without actually using it?

bennyz replied:
leftover from a previous attempt, remove it

bennyz committed:
Minor issues

- Use empty struct instead of bool in util.Difference
- Remove unsued cachedRepos

bennyz committed:
Group global variables together
